### ISSUE TITLE
Bug: [no-unnecessary-type-assertion] (regression) false positives on template literals with no expressions

### ISSUE THREAD

**Issue by abrahamguo** [typescript-eslint/typescript-eslint#8819]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link
https://typescript-eslint.io/play/#ts=5.4.3&fileType=.tsx&code=DYUwLgBAhhC8EAMoOgZwgYwPYDtVgG4AoUSAIzggHIyq1Nd8Cg&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0UOdA7RI4MAF8QOoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
let a = `a` as const;
let b = 'b' as const;
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unnecessary-type-assertion": "error"
  },
};
```

### Expected Result
No lint errors should be reported, as the type assertions are meaningful.

### Actual Result
The type assertion on the backtick strings is reported as unnecessary, while the quoted string is not reported.

### Additional Info
A regression from #8740
